zephyr/test/vmh.c

```diff
@@ -17,8 +18,258 @@
+	if (expect_success) {
+		zassert_not_null(heap,
+		"Heap initialization expected to succeed but failed");
+		}
+	else
+		zassert_is_null(heap, "Heap initialization expected to fail but succeeded");
+
+	if (heap) {
+		int ret = vmh_free_heap(heap);
+
+		zassert_equal(ret, 0, "Failed to free heap");
+	}
+}
+
+/* Test for vmh_alloc and vmh_free */
+static void test_vmh_alloc_free_no_check(struct vmh_heap *heap,
+					uint32_t alloc_size,
+					bool expect_success)
+{
+	void *ptr = vmh_alloc(heap, alloc_size);
+
+	if (expect_success)
+		zassert_not_null(ptr, "Allocation expected to succeed but failed");
+	else
+		zassert_is_null(ptr, "Allocation expected to fail but succeeded");
+
+	if (ptr) {
+		int ret = vmh_free(heap, ptr);
+
+		zassert_equal(ret, 0, "Failed to free allocated memory");
+	}
+}
+
+/* Test function for vmh_alloc and vmh_free with memory read/write */
+static void test_vmh_alloc_free_check(struct vmh_heap *heap,
+					uint32_t alloc_size,
+					bool expect_success)
+{
+	void *ptr = vmh_alloc(heap, alloc_size);
+
+	if (expect_success)
+		zassert_not_null(ptr, "Allocation expected to succeed but failed");
+	else {
+		zassert_is_null(ptr, "Allocation expected to fail but succeeded");
+		return;
+	}
+
+	if (ptr) {
+		/* Write test pattern to the allocated memory and verify it */
+		memset(ptr, 0xAA, alloc_size);
+
+		uint8_t start_val = *((uint8_t *)ptr);
+
+		zassert_equal(start_val, 0xAA,
+		"Memory content verification failed at the start");
+
+		uint8_t end_val = *((uint8_t *)ptr + alloc_size - 1);
+
+		zassert_equal(end_val, 0xAA,
+		"Memory content verification failed at the end");
+
+		if (alloc_size > 2) {
+			uint32_t middle_pos = alloc_size / 2;
+			uint8_t middle_val = *((uint8_t *)ptr + middle_pos);
+
+			zassert_equal(middle_val, 0xAA,
+			"Memory content verification failed in the middle");
+	}
+
+		int ret = vmh_free(heap, ptr);
+
+		zassert_equal(ret, 0, "Failed to free allocated memory");
+	}
+}
+
+/* Test function for multiple allocations on the same heap with read/write */
+static void test_vmh_multiple_allocs(struct vmh_heap *heap, int num_allocs,
+					uint32_t min_alloc_size,
+					uint32_t max_alloc_size)
+{
+	void *ptrs[num_allocs];
+	uint32_t alloc_size;
+	bool success;
+	int ret;
+
+	/* Perform multiple allocations */
+	for (int i = 0; i < num_allocs; i++) {
+		/* Generate a random allocation size between min_alloc_size and max_alloc_size */
+		alloc_size = min_alloc_size +
+		k_cycle_get_32() % (max_alloc_size - min_alloc_size + 1);
+
+		ptrs[i] = vmh_alloc(heap, alloc_size);
+
+		zassert_true(ptrs[i] != NULL,
+		"Allocation of size %u expected to succeed but failed",
+		alloc_size);
+
+	if (ptrs[i] != NULL) {
+		LOG_INF("Test allocation size: %d", alloc_size);
+
+		memset(ptrs[i], 0xAA, alloc_size);
+
+
+		uint8_t *byte_ptr = (uint8_t *)ptrs[i];
+
+		for (uint32_t j = 0; j < alloc_size; j++) {
+			zassert_equal(byte_ptr[j], 0xAA,
+			"Memory content verification failed at position %u", j);
```

Comment:
I *really* don't think we should check every byte in every allocation in multiple allocations in multiple tests in multiple heaps

Reply:
The problem I am trying to catch here is that allocation worked but physical pages were not mapped correctly. I am afraid I do not see any other option to be sure.
If our discussion ends with CONFIG_SOF_BOOT_TEST being time constrained we can think about timing it down. Let's come back to it after we clear that up in the comments above.

Reply:
I think the comments is possible this could be faster and more optimal with words instead of bytes ? or even use memcmp() now that we have a C library in Zephyr

Reply:
I will try to rewrite the check to use bigger data, however, memcmp compares memory regions so it is not applicable if I have no buffer with the pattern.